<a href="https://colab.research.google.com/github/gracewindheim/gmwDataSci/blob/master/q2hw8.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Question 2

Use Keras and Tensor Flow on colab to fit a prediction model for the fashion mnist data. Try 5 versions of the model with different numbers of hidden layers and dropout (of your choosing). Compare the test set accuracy of each of the models.

In [0]:
!pip install -U tensorflow_datasets
import tensorflow as tf
# Import TensorFlow Datasets
import tensorflow_datasets as tfds
tfds.disable_progress_bar()

# Helper libraries
import math
import numpy as np
import matplotlib.pyplot as plt

import logging
logger = tf.get_logger()
logger.setLevel(logging.ERROR)

     |████████████████████████████████| 3.3MB 3.2MB/s 
  Found existing installation: tensorflow-datasets 2.1.0
    Uninstalling tensorflow-datasets-2.1.0:
      Successfully uninstalled tensorflow-datasets-2.1.0


## Preprocess data

In [0]:
dataset, metadata = tfds.load('fashion_mnist', as_supervised=True, with_info=True)
train_dataset, test_dataset = dataset['train'], dataset['test']

class_names = ['T-shirt/top', 'Trouser', 'Pullover', 'Dress', 'Coat',
               'Sandal',      'Shirt',   'Sneaker',  'Bag',   'Ankle boot']

# Preprocessing data
def normalize(images, labels):
  images = tf.cast(images, tf.float32)
  images /= 255
  return images, labels

# The map function applies the normalize function to each element in the train
# and test datasets
train_dataset =  train_dataset.map(normalize)
test_dataset  =  test_dataset.map(normalize)

# The first time you use the dataset, the images will be loaded from disk
# Caching will keep them in memory, making training faster
train_dataset =  train_dataset.cache()
test_dataset  =  test_dataset.cache()

Shuffling and writing examples to /root/tensorflow_datasets/fashion_mnist/3.0.1.incomplete0XQHXM/fashion_mnist-train.tfrecord
Shuffling and writing examples to /root/tensorflow_datasets/fashion_mnist/3.0.1.incomplete0XQHXM/fashion_mnist-test.tfrecord
Dataset fashion_mnist downloaded and prepared to /root/tensorflow_datasets/fashion_mnist/3.0.1. Subsequent calls will reuse this data.


In [0]:
num_train_examples = metadata.splits['train'].num_examples
num_test_examples = metadata.splits['test'].num_examples
print("Number of training examples: {}".format(num_train_examples))
print("Number of test examples:     {}".format(num_test_examples))

Number of training examples: 60000
Number of test examples:     10000


## Version 1
1 hidden layer, 10% dropout

In [0]:
model1 = tf.keras.Sequential([
    tf.keras.layers.Flatten(input_shape=(28, 28, 1)),
    tf.keras.layers.Dense(128, activation=tf.nn.relu),
    tf.keras.layers.Dropout(0.1),
    tf.keras.layers.Dense(10)
])

model1.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])

BATCH_SIZE = 32
train_dataset1 = train_dataset.cache().repeat().shuffle(num_train_examples).batch(BATCH_SIZE)
test_dataset1 = test_dataset.cache().batch(BATCH_SIZE)
model1.fit(train_dataset1, epochs=5, steps_per_epoch=math.ceil(num_train_examples/BATCH_SIZE))

test_loss, test_accuracy = model1.evaluate(test_dataset1, steps=math.ceil(num_test_examples/32))
print('Accuracy on test dataset:', test_accuracy)

Epoch 1/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.5053 - accuracy: 0.8218
Epoch 2/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.3856 - accuracy: 0.8607
Epoch 3/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.3461 - accuracy: 0.8731
Epoch 4/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.3208 - accuracy: 0.8812
Epoch 5/5
313/313 [==============================] - 1s 4ms/step - loss: 0.3480 - accuracy: 0.8757
Accuracy on test dataset: 0.8756999969482422


## Version 2
3 hidden layers, 10% dropout

In [0]:
model2 = tf.keras.Sequential([
    tf.keras.layers.Flatten(input_shape=(28, 28, 1)),
    tf.keras.layers.Dense(128, activation=tf.nn.relu),
    tf.keras.layers.Dropout(0.1),
    tf.keras.layers.Dense(128, activation=tf.nn.relu),
    tf.keras.layers.Dropout(0.1),
    tf.keras.layers.Dense(128, activation=tf.nn.relu),
    tf.keras.layers.Dropout(0.1),
    tf.keras.layers.Dense(10)
])

model2.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])

BATCH_SIZE = 32
train_dataset2 = train_dataset.cache().repeat().shuffle(num_train_examples).batch(BATCH_SIZE)
test_dataset2 = test_dataset.cache().batch(BATCH_SIZE)
model2.fit(train_dataset2, epochs=5, steps_per_epoch=math.ceil(num_train_examples/BATCH_SIZE))

test_loss, test_accuracy = model2.evaluate(test_dataset2, steps=math.ceil(num_test_examples/32))
print('Accuracy on test dataset:', test_accuracy)

Epoch 1/5
1875/1875 [==============================] - 5s 3ms/step - loss: 0.5303 - accuracy: 0.8084
Epoch 2/5
1875/1875 [==============================] - 5s 3ms/step - loss: 0.4010 - accuracy: 0.8537
Epoch 3/5
1875/1875 [==============================] - 5s 2ms/step - loss: 0.3664 - accuracy: 0.8669
Epoch 4/5
1875/1875 [==============================] - 5s 3ms/step - loss: 0.3471 - accuracy: 0.8711
Epoch 5/5
313/313 [==============================] - 0s 1ms/step - loss: 0.3516 - accuracy: 0.8728
Accuracy on test dataset: 0.8727999925613403


## Version 3
5 hidden layers, 10% dropout

In [0]:
model3 = tf.keras.Sequential([
    tf.keras.layers.Flatten(input_shape=(28, 28, 1)),
    tf.keras.layers.Dense(128, activation=tf.nn.relu),
    tf.keras.layers.Dropout(0.1),
    tf.keras.layers.Dense(128, activation=tf.nn.relu),
    tf.keras.layers.Dropout(0.1),
    tf.keras.layers.Dense(128, activation=tf.nn.relu),
    tf.keras.layers.Dropout(0.1),
    tf.keras.layers.Dense(128, activation=tf.nn.relu),
    tf.keras.layers.Dropout(0.1),
    tf.keras.layers.Dense(128, activation=tf.nn.relu),
    tf.keras.layers.Dropout(0.1),
    tf.keras.layers.Dense(10)
])

model3.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])

BATCH_SIZE = 32
train_dataset3 = train_dataset.cache().repeat().shuffle(num_train_examples).batch(BATCH_SIZE)
test_dataset3 = test_dataset.cache().batch(BATCH_SIZE)
model3.fit(train_dataset3, epochs=5, steps_per_epoch=math.ceil(num_train_examples/BATCH_SIZE))

test_loss, test_accuracy = model3.evaluate(test_dataset3, steps=math.ceil(num_test_examples/32))
print('Accuracy on test dataset:', test_accuracy)

Epoch 1/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.5769 - accuracy: 0.7933
Epoch 2/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.4189 - accuracy: 0.8496
Epoch 3/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.3904 - accuracy: 0.8605
Epoch 4/5
1875/1875 [==============================] - 5s 3ms/step - loss: 0.3747 - accuracy: 0.8662
Epoch 5/5
313/313 [==============================] - 0s 1ms/step - loss: 0.3667 - accuracy: 0.8669
Accuracy on test dataset: 0.8669000267982483


## Version 4
2 hidden layers, 40% dropout

In [0]:
model4 = tf.keras.Sequential([
    tf.keras.layers.Flatten(input_shape=(28, 28, 1)),
    tf.keras.layers.Dense(128, activation=tf.nn.relu),
    tf.keras.layers.Dropout(0.4),
    tf.keras.layers.Dense(128, activation=tf.nn.relu),
    tf.keras.layers.Dropout(0.4),
    tf.keras.layers.Dense(10)
])

model4.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])

BATCH_SIZE = 32
train_dataset4 = train_dataset.cache().repeat().shuffle(num_train_examples).batch(BATCH_SIZE)
test_dataset4 = test_dataset.cache().batch(BATCH_SIZE)
model4.fit(train_dataset4, epochs=5, steps_per_epoch=math.ceil(num_train_examples/BATCH_SIZE))

test_loss, test_accuracy = model4.evaluate(test_dataset4, steps=math.ceil(num_test_examples/32))
print('Accuracy on test dataset:', test_accuracy)

Epoch 1/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.6476 - accuracy: 0.7659
Epoch 2/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.4870 - accuracy: 0.8251
Epoch 3/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.4468 - accuracy: 0.8396
Epoch 4/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.4272 - accuracy: 0.8453
Epoch 5/5
313/313 [==============================] - 0s 1ms/step - loss: 0.3827 - accuracy: 0.8604
Accuracy on test dataset: 0.8604000210762024


## Version 5
2 hidden layer, 80% dropout

In [0]:
model5 = tf.keras.Sequential([
    tf.keras.layers.Flatten(input_shape=(28, 28, 1)),
    tf.keras.layers.Dense(128, activation=tf.nn.relu),
    tf.keras.layers.Dropout(0.8),
    tf.keras.layers.Dense(128, activation=tf.nn.relu),
    tf.keras.layers.Dropout(0.8),
    tf.keras.layers.Dense(10)
])

model5.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])

BATCH_SIZE = 32
train_dataset5 = train_dataset.cache().repeat().shuffle(num_train_examples).batch(BATCH_SIZE)
test_dataset5 = test_dataset.cache().batch(BATCH_SIZE)
model5.fit(train_dataset5, epochs=5, steps_per_epoch=math.ceil(num_train_examples/BATCH_SIZE))

test_loss, test_accuracy = model5.evaluate(test_dataset5, steps=math.ceil(num_test_examples/32))
print('Accuracy on test dataset:', test_accuracy)

Epoch 1/5
1875/1875 [==============================] - 4s 2ms/step - loss: 1.5940 - accuracy: 0.3966
Epoch 2/5
1875/1875 [==============================] - 4s 2ms/step - loss: 1.2447 - accuracy: 0.5121
Epoch 3/5
1875/1875 [==============================] - 4s 2ms/step - loss: 1.1617 - accuracy: 0.5430
Epoch 4/5
1875/1875 [==============================] - 4s 2ms/step - loss: 1.1242 - accuracy: 0.5587
Epoch 5/5
313/313 [==============================] - 0s 2ms/step - loss: 0.7824 - accuracy: 0.7116
Accuracy on test dataset: 0.7116000056266785


## Comparison

Version 1 (1 hidden layer, 10% dropout). Accuracy = 87.6%

Version 2 (3 hidden layers, 10% dropout). Accuracy = 87.3%

Version 3 (5 hidden layers, 10% dropout). Accuracy = 86.7%

Version 4 (2 hidden layers, 40% dropout). Accuracy = 86.0%

Version 5 (2 hidden layers, 80% dropout). Accuracy = 71.2%


Seems to perform better with fewer layers and fewer dropout.